File: white-hole/src/Operations/UserOperations.hs
{-# LANGUAGE OverloadedStrings #-}
module Operations.UserOperations where

import Database.PostgreSQL.Simple
import Entities.User
import qualified Entities.Movie as M
import qualified Entities.Rating as R
import qualified Entities.Serie as S


createUser :: Connection -> User -> IO Bool
createUser conn user = do
    exists <- userAlreadyExists conn $ email user
    if not exists then do
        ok <- execute conn "INSERT INTO users (email, passwd, nome, sobrenome) values (?,?,?,?);" (email user, password user, nome user, sobrenome user)
        return True
    else return False


getUserByEmail :: Connection -> String -> IO [User]
getUserByEmail conn userEmail = do
    let q = "SELECT * FROM users WHERE email = ?;"
    query conn q [userEmail] :: IO [User]


userAlreadyExists :: Connection -> String -> IO Bool
userAlreadyExists conn userEmail = do
    result <- getUserByEmail conn userEmail
    return (not (null result))


authenticate :: Connection -> String -> String -> IO Bool
authenticate conn userEmail userPassword = do
    result <- getUserByEmail conn userEmail
    if not (null result) then do
        return (password (head result) == userPassword)
    else
        return False


addToWatchLaterList :: Connection -> User -> M.Movie -> IO Bool
addToWatchLaterList conn user movie = do
    userExists <- userAlreadyExists conn $ email user
    if userExists then do
        execute conn "INSERT INTO watchlaterlist values (?,?)" (email user, M.movieId movie)
        return True
    else do
        return False

File: white-hole-pl/src/Operations/user_operations.pl
:- module(user_operations, [createUser/6, getUserByEmail/3, userAlreadyExists/3, authenticate/4, getUsersByEmail/4]).
:- use_module('./util.pl').
:- use_module('../Db/dbOperations.pl').


createUser(Connection, Email, Senha, Nome, Sobrenome, Confirmacao):-
    userAlreadyExists(Connection, Email, Fstconf),
    ( Fstconf =:= 0 ->
        db_parameterized_query_no_return(
            Connection, 
            "INSERT INTO users (email, passwd, nome, sobrenome) values ('%w','%w','%w','%w');",
            [Email, Senha, Nome, Sobrenome]),
        Confirmacao is 1;
        Confirmacao is 0).


getUserByEmail(Connection, Email, User):-
    Q = "SELECT * FROM users WHERE email = '%w'",
    db_parameterized_query(Connection, Q, [Email], User).


userAlreadyExists(Connection, Email, Confirmacao):-
    getUserByEmail(Connection, Email, User),
    length(User, Confirmacao).


authenticate(Connection, Email, Senha, Confirmacao):-
    getUserByEmail(Connection, Email, User),
    length(User, L),
    (L =:= 0 ->
        Confirmacao is 0;
        User = [ row(_, SenhaReal, _, _) | _ ],
        (Senha = SenhaReal -> Confirmacao is 1; Confirmacao is 0)
    ).


getUsersByEmail(Connection, [ Email | T ], UsersTemp, Users):-
    length(T, L),
    (L > 0 ->
        getUserByEmail(Connection, Email, User),
        getUsersByEmail(Connection, T, [ User | UsersTemp ], Users) ;
        getUserByEmail(Connection, Email, User),
        reverse([ User | UsersTemp ], Users)
    ).